xpdan/data_reduction_core.py

```diff
@@ -0,0 +1,576 @@
+        if el is not None:
+            # truncate string length
+            if len(el) > 12:
+                value = el[:12]
+            # clear space
+            feature = _clean_info(el)
+            feature_list.append(feature)
+        else:
+            pass
+    # get data fields
+    for key in data_fields:
+        val = event['data'].get(key, None)
+        if val is not None:
+            feature = "{}".format(val)
+            feature_list.append(feature)
+        else:
+            pass
+    # get uid
+    feature_list.append(uid)
+    return "_".join(feature_list)
+
+
+def _file_name(event, event_timestamp, ind):
+    """ priviate method operates on event level """
+    f_name = _feature_gen(event)
+    f_name = '_'.join([_timestampstr(event_timestamp),
+                       f_name])
+    f_name = '{}_{:04d}.tif'.format(f_name, ind)
+    return f_name
+
+
+# init
+
+""" analysis function operates at header level """
+
+
+def _prepare_header_list(headers):
+    if not isinstance(headers, list):
+        # still do it in two steps, easier to read
+        header_list = list()
+        header_list.append(headers)
+    else:
+        header_list = headers
+    return header_list
+
+
+def _load_config(header, config_base='', calib_config_name=''):
+
+    try:
+        with open(os.path.join(config_base, calib_config_name)) as f:
+            config_dict = yaml.load(f)
+    except FileNotFoundError:
+        config_dict = header.start.get('calibration_md', None)
+        if config_dict is None:
+            # back support
+            config_dict = header.start.get('sc_calibration_md', None)
+
+    return config_dict
+
+
+def _npt_cal(config_dict, total_shape=(2048, 2048)):
+    """ config_dict should be a PyFAI calibration dict """
+    x_0, y_0 = (config_dict['centerX'], config_dict['centerY'])
+    center_len = np.hypot(x_0, y_0)
+    # FIXME : use hardwired shape now, use direct info later
+    x_total, y_total = total_shape
+    total_len = np.hypot(x_total, y_total)
+    # FIXME : use the longest diagonal distance. Optimal value might have
+    # to do with grid of Fourier transform. Need to revisit it later
+    dist = max(total_len, total_len - center_len)
+    return dist
+
+
+def integrate_and_save(headers, *, db, save_dir,
+                       path_append_keys='sample_name',
+                       dark_sub_bool=True,
+                       polarization_factor=0.99,
+                       mask_setting='default',
+                       mask_dict=None,
+                       save_image=True,
+                       config_dict=None,
+                       image_data_key='pe1_image',
+                       **kwargs):
+    """ integrate and save dark subtracted images for given list of headers
+
+    Parameters
+    ----------
+    headers : list
+        a list of databroker.header objects
+    db: databroker.broker.Broker instance
+        The databroker which has all the data
```

Comment:
must be specified as db=<databroker object> in the argument (keyword only argument)